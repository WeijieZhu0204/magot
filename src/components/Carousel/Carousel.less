@import '../../variable.less';

.@{prefix}-carousel {
  position: relative;
  .box();
  .box-sizing();
}

.@{prefix}-carousel-cardbox {
  position: relative;
  overflow: hidden;
}

.@{prefix}-carousel-cards {
  position: relative;
  list-style: none;
  .box();
  .box-sizing();
  .clearfix();

  &.slide {
    transform: translate3d(0, 0, 0);
    .transition(transform; cubic-bezier(0.86, 0, 0.07, 1) ; 800ms;);
  }

  &.fade {
    .@{prefix}-carousel-card {
      opacity: 0;
      pointer-events: none;
      .transition(opacity; cubic-bezier(0.86, 0, 0.07, 1) ; 500ms);
      &.active {
        opacity: 1;
        pointer-events: auto;
      }
    }
  }
}

.@{prefix}-carousel-card {
  float: left;
  position: relative;
  left: 0;
  height: 100%;
}

.@{prefix}-carousel-dots {
  position: absolute;
  bottom: @carousel-dots-bottom;
  width: 100%;
  height: @carousel-dots-height;
  text-align: center;
  list-style: none;
  z-index: 2;
  .box();
}

.@{prefix}-carousel-dot {
  display: inline-block;
  margin: 0 @carousel-dot-gap;
  vertical-align: top;

  & > button {
    display: block;
    padding: 0;
    width: @carousel-dot-width;
    height: @carousel-dots-height;
    background: none;
    border: none;
    outline: none;
    color: transparent;
    background-color: @carousel-dot-background-color;
    border-radius: 1px;
    cursor: pointer;
    .transition();
  }

  &.active {
    & > button {
      width: @carousel-dot-width-active;
      background-color: @carousel-dot-background-color-active;
    }
  }
}
