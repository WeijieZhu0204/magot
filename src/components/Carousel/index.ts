export { default, CarouselProps, CarouselCardProps } from './Carousel';
