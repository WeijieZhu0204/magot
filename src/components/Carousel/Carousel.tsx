import React from 'react';
import cx from 'classnames';

import './Carousel.less';

import * as component from '../component';

export interface CarouselProps
  extends component.BaseComponent,
    component.NestedComponent {
  /**
   * 是否自动切换
   * @default false
   */
  autoPlay?: boolean;

  /**
   * 自动切换的间隔时长，单位毫秒
   * @default 3000
   */
  interval?: number;

  /**
   * 是否显示指示点
   * @default true
   */
  showDots?: boolean;

  /**
   * 动画效果，slide-滑动, fade-淡变
   * @default slide
   */
  effect?: 'slide' | 'fade';

  children?: React.FunctionComponentElement<CarouselCardProps>[];

  /**
   * 切换后的事件回调函数
   */
  onChange?: (current: number, old: number) => void;
}

interface CarouselCardsProps extends Pick<CarouselProps, 'children'> {
  prefix: string;
  width: number;
  current: number;
  effect: 'slide' | 'fade';
}

export interface CarouselCardProps
  extends component.BaseComponent,
    component.NestedComponent {}

interface CarouselDotsProps {
  count: number;
  prefix: string;
  current: number;
  onSwitch: (index: number) => void;
}

const defaultProps: Partial<CarouselProps> = {
  autoPlay: false,
  interval: 3000,
  showDots: true,
  effect: 'slide',
};

function CarouselCards(props: CarouselCardsProps) {
  const count = React.Children.count(props.children);
  const totalWidth = props.width * count;

  let cards = null;
  if (props.children) {
    cards = React.Children.toArray(props.children).map((card, index) => {
      const isActived = index === props.current;
      return React.cloneElement(card, {
        className: cx(card.props.className, { active: isActived }),
        style: {
          ...card.props.style,
          width: props.width,
          display: !props.width ? 'none' : '',
          left: props.effect === 'fade' ? -index * props.width : 0,
        },
      });
    });
  }

  const style: Record<string, any> = { width: totalWidth };
  if (props.effect === 'slide') {
    const x = -props.current * props.width;
    style.transform = `translate3d(${x}px, 0, 0)`;
  }

  return (
    <ul className={cx(props.prefix + '-cards', props.effect)} style={style}>
      {cards}
    </ul>
  );
}

function CarouselCard(props: CarouselCardProps) {
  const cls = component.getComponentClasses('carousel-card', props);
  return (
    <li className={cls} style={props.style}>
      {props.children}
    </li>
  );
}

function CarouselDots(props: CarouselDotsProps) {
  return (
    <ul className={props.prefix + '-dots'}>
      {Array.from({ length: props.count }).map((_v, i) => {
        const isActived = props.current === i;
        return (
          <li
            className={cx(props.prefix + '-dot', { active: isActived })}
            key={i}>
            <button type="button" onClick={() => props.onSwitch(i)} />
          </li>
        );
      })}
    </ul>
  );
}

function Carousel(props: CarouselProps) {
  const count = React.Children.count(props.children);
  const carouselRef = React.useRef<HTMLDivElement | null>(null);
  const [carouselWidth, setCarouselWidth] = React.useState(0);
  const [currentIndex, setCurrentIndex] = React.useState(0);

  const handleDotSwitch = (index: number) => {
    setCurrentIndex(index);
    props.onChange && props.onChange(index, currentIndex);
  };

  React.useEffect(() => {
    if (carouselRef.current) {
      const width = carouselRef.current.offsetWidth;
      if (width !== carouselWidth) setCarouselWidth(width);
    }
    let timer = 0;
    if (props.autoPlay && (props.interval || 0) > 0) {
      timer = window.setTimeout(() => {
        let index = currentIndex + 1;
        if (index >= count) index = 0;
        handleDotSwitch(index);
      }, props.interval);
    }
    return () => window.clearTimeout(timer);
  });

  const type = 'carousel';
  const prefix = component.getComponentPrefix(type);
  const cls = component.getComponentClasses(type, props);

  return (
    <div ref={carouselRef} className={cls}>
      <div className={prefix + '-cardbox'}>
        <CarouselCards
          prefix={prefix}
          width={carouselWidth}
          current={currentIndex}
          effect={props.effect!}>
          {props.children}
        </CarouselCards>
      </div>
      {props.showDots && (
        <CarouselDots
          count={count}
          prefix={prefix}
          current={currentIndex}
          onSwitch={handleDotSwitch}
        />
      )}
    </div>
  );
}

Carousel.defaultProps = defaultProps;
Carousel.Card = CarouselCard;

export default Carousel;
