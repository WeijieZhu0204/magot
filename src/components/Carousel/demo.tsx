import React from 'react';
import Carousel from './Carousel';
import { Link } from 'react-router-dom';

function CarouselDemo() {
  const cardStyle = {
    height: 240,
    backgroundColor: '#20a0ff',
    color: '#fff',
    fontSize: 48,
  };
  return (
    <>
      <Link to="/" className="demo-goback">
        返回
      </Link>
      <div className="demo-box">
        <Carousel onChange={console.log}>
          <Carousel.Card style={cardStyle}>
            <h3 className="demo-center-block">1</h3>
          </Carousel.Card>
          <Carousel.Card style={cardStyle}>
            <h3 className="demo-center-block">2</h3>
          </Carousel.Card>
          <Carousel.Card style={cardStyle}>
            <h3 className="demo-center-block">3</h3>
          </Carousel.Card>
          <Carousel.Card style={cardStyle}>
            <h3 className="demo-center-block">4</h3>
          </Carousel.Card>
        </Carousel>
      </div>
      <div className="demo-box">
        <Carousel autoPlay={true}>
          <Carousel.Card style={cardStyle}>
            <h3 className="demo-center-block">1</h3>
          </Carousel.Card>
          <Carousel.Card style={cardStyle}>
            <h3 className="demo-center-block">2</h3>
          </Carousel.Card>
          <Carousel.Card style={cardStyle}>
            <h3 className="demo-center-block">3</h3>
          </Carousel.Card>
          <Carousel.Card style={cardStyle}>
            <h3 className="demo-center-block">4</h3>
          </Carousel.Card>
        </Carousel>
      </div>
      <div className="demo-box">
        <Carousel autoPlay={true} effect="fade">
          <Carousel.Card style={cardStyle}>
            <h3 className="demo-center-block">1</h3>
          </Carousel.Card>
          <Carousel.Card style={cardStyle}>
            <h3 className="demo-center-block">2</h3>
          </Carousel.Card>
          <Carousel.Card style={cardStyle}>
            <h3 className="demo-center-block">3</h3>
          </Carousel.Card>
          <Carousel.Card style={cardStyle}>
            <h3 className="demo-center-block">4</h3>
          </Carousel.Card>
        </Carousel>
      </div>
    </>
  );
}

export default CarouselDemo;
