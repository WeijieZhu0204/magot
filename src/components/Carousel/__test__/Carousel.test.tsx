import React from 'react';
import { render, mount } from 'enzyme';
import Carousel from '../Carousel';
import { injectStyleTestSuites } from '../../../../test/shared.test';

describe('Carousel', () => {
  it('should renders dom correctly', () => {
    const wrapper = render(<Carousel />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should renders have children', () => {
    const wrapper = render(
      <Carousel>
        <Carousel.Card>
          <h3>1</h3>
        </Carousel.Card>
        <Carousel.Card>
          <h3>2</h3>
        </Carousel.Card>
        <Carousel.Card>
          <h3>3</h3>
        </Carousel.Card>
      </Carousel>
    );
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('.mgt-carousel-card')).toHaveLength(3);
  });

  it('should renders `autoPlay` when pass in', () => {
    const wrapper = render(
      <Carousel autoPlay={true}>
        <Carousel.Card>
          <h3>1</h3>
        </Carousel.Card>
        <Carousel.Card>
          <h3>2</h3>
        </Carousel.Card>
        <Carousel.Card>
          <h3>3</h3>
        </Carousel.Card>
      </Carousel>
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should renders `showDots` is false', () => {
    const wrapper = render(
      <Carousel showDots={false}>
        <Carousel.Card>
          <h3>1</h3>
        </Carousel.Card>
        <Carousel.Card>
          <h3>2</h3>
        </Carousel.Card>
        <Carousel.Card>
          <h3>3</h3>
        </Carousel.Card>
      </Carousel>
    );
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('.mgt-carousel-dots')).toHaveLength(0);
  });

  it('should renders `effect` when pass in', () => {
    const wrapper = render(
      <Carousel effect="fade">
        <Carousel.Card>
          <h3>1</h3>
        </Carousel.Card>
        <Carousel.Card>
          <h3>2</h3>
        </Carousel.Card>
        <Carousel.Card>
          <h3>3</h3>
        </Carousel.Card>
      </Carousel>
    );
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('.mgt-carousel-cards.fade')).toHaveLength(1);
  });

  it('should simulates change event', () => {
    const handleChange = jest.fn();
    const wrapper = mount(
      <Carousel onChange={handleChange}>
        <Carousel.Card>
          <h3>1</h3>
        </Carousel.Card>
        <Carousel.Card>
          <h3>2</h3>
        </Carousel.Card>
        <Carousel.Card>
          <h3>3</h3>
        </Carousel.Card>
      </Carousel>
    );
    wrapper
      .find('.mgt-carousel-dot')
      .at(2)
      .find('button')
      .simulate('click');
    expect(handleChange).toBeCalledWith(2, 0);
  });

  injectStyleTestSuites(Carousel);
});
