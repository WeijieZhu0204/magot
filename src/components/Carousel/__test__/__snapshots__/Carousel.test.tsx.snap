// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Carousel should renders \`autoPlay\` when pass in 1`] = `
<div
  class="mgt-carousel"
>
  <div
    class="mgt-carousel-cardbox"
  >
    <ul
      class="mgt-carousel-cards slide"
      style="width:0;transform:translate3d(0px, 0, 0)"
    >
      <li
        class="mgt-carousel-card active"
        style="width:0;display:none;left:0"
      >
        <h3>
          1
        </h3>
      </li>
      <li
        class="mgt-carousel-card"
        style="width:0;display:none;left:0"
      >
        <h3>
          2
        </h3>
      </li>
      <li
        class="mgt-carousel-card"
        style="width:0;display:none;left:0"
      >
        <h3>
          3
        </h3>
      </li>
    </ul>
  </div>
  <ul
    class="mgt-carousel-dots"
  >
    <li
      class="mgt-carousel-dot active"
    >
      <button
        type="button"
      />
    </li>
    <li
      class="mgt-carousel-dot"
    >
      <button
        type="button"
      />
    </li>
    <li
      class="mgt-carousel-dot"
    >
      <button
        type="button"
      />
    </li>
  </ul>
</div>
`;

exports[`Carousel should renders \`effect\` when pass in 1`] = `
<div
  class="mgt-carousel"
>
  <div
    class="mgt-carousel-cardbox"
  >
    <ul
      class="mgt-carousel-cards fade"
      style="width:0"
    >
      <li
        class="mgt-carousel-card active"
        style="width:0;display:none;left:0"
      >
        <h3>
          1
        </h3>
      </li>
      <li
        class="mgt-carousel-card"
        style="width:0;display:none;left:0"
      >
        <h3>
          2
        </h3>
      </li>
      <li
        class="mgt-carousel-card"
        style="width:0;display:none;left:0"
      >
        <h3>
          3
        </h3>
      </li>
    </ul>
  </div>
  <ul
    class="mgt-carousel-dots"
  >
    <li
      class="mgt-carousel-dot active"
    >
      <button
        type="button"
      />
    </li>
    <li
      class="mgt-carousel-dot"
    >
      <button
        type="button"
      />
    </li>
    <li
      class="mgt-carousel-dot"
    >
      <button
        type="button"
      />
    </li>
  </ul>
</div>
`;

exports[`Carousel should renders \`showDots\` is false 1`] = `
<div
  class="mgt-carousel"
>
  <div
    class="mgt-carousel-cardbox"
  >
    <ul
      class="mgt-carousel-cards slide"
      style="width:0;transform:translate3d(0px, 0, 0)"
    >
      <li
        class="mgt-carousel-card active"
        style="width:0;display:none;left:0"
      >
        <h3>
          1
        </h3>
      </li>
      <li
        class="mgt-carousel-card"
        style="width:0;display:none;left:0"
      >
        <h3>
          2
        </h3>
      </li>
      <li
        class="mgt-carousel-card"
        style="width:0;display:none;left:0"
      >
        <h3>
          3
        </h3>
      </li>
    </ul>
  </div>
</div>
`;

exports[`Carousel should renders custom className 1`] = `
<div
  class="mgt-carousel jest-test-class"
>
  <div
    class="mgt-carousel-cardbox"
  >
    <ul
      class="mgt-carousel-cards slide"
      style="width:0;transform:translate3d(0px, 0, 0)"
    />
  </div>
  <ul
    class="mgt-carousel-dots"
  />
</div>
`;

exports[`Carousel should renders custom styles 1`] = `
<div
  class="mgt-carousel"
  style="color:#333;background-color:#f0f0f0"
>
  <div
    class="mgt-carousel-cardbox"
  >
    <ul
      class="mgt-carousel-cards slide"
      style="width:0;transform:translate3d(0px, 0, 0)"
    />
  </div>
  <ul
    class="mgt-carousel-dots"
  />
</div>
`;

exports[`Carousel should renders dom correctly 1`] = `
<div
  class="mgt-carousel"
>
  <div
    class="mgt-carousel-cardbox"
  >
    <ul
      class="mgt-carousel-cards slide"
      style="width:0;transform:translate3d(0px, 0, 0)"
    />
  </div>
  <ul
    class="mgt-carousel-dots"
  />
</div>
`;

exports[`Carousel should renders have children 1`] = `
<div
  class="mgt-carousel"
>
  <div
    class="mgt-carousel-cardbox"
  >
    <ul
      class="mgt-carousel-cards slide"
      style="width:0;transform:translate3d(0px, 0, 0)"
    >
      <li
        class="mgt-carousel-card active"
        style="width:0;display:none;left:0"
      >
        <h3>
          1
        </h3>
      </li>
      <li
        class="mgt-carousel-card"
        style="width:0;display:none;left:0"
      >
        <h3>
          2
        </h3>
      </li>
      <li
        class="mgt-carousel-card"
        style="width:0;display:none;left:0"
      >
        <h3>
          3
        </h3>
      </li>
    </ul>
  </div>
  <ul
    class="mgt-carousel-dots"
  >
    <li
      class="mgt-carousel-dot active"
    >
      <button
        type="button"
      />
    </li>
    <li
      class="mgt-carousel-dot"
    >
      <button
        type="button"
      />
    </li>
    <li
      class="mgt-carousel-dot"
    >
      <button
        type="button"
      />
    </li>
  </ul>
</div>
`;
